kernel/hashlib.h

```diff
@@ -441,6 +455,26 @@ class dict
 		return std::pair<iterator, bool>(iterator(this, i), true);
 	}
 
+	std::pair<iterator, bool> insert(std::pair<K, T> &&rvalue)
+	{
+		int hash = do_hash(rvalue.first);
+		int i = do_lookup(rvalue.first, hash);
+		if (i >= 0)
+			return std::pair<iterator, bool>(iterator(this, i), false);
+		i = do_insert(std::forward<std::pair<K, T>>(rvalue), hash);
+		return std::pair<iterator, bool>(iterator(this, i), true);
+	}
+
+	std::pair<iterator, bool> insert(K const &key, T &&rvalue)
```

Comment:
In keeping with STL semantics, shouldn't this be an `emplace()`?

Reply:
Maybe... I actually had played with the `emplace()` name originally.  Technically I think a proper `emplace()` also uses some fancy template varargs stuff that I am not familiar with.  Its usage does *look* the same, though.

TBH my main reason for including this specific function wasn't to make a proper `emplace()` but to avoid an ugly explicit `std::make_pair()` in `passes/cmds/scatter.cc`.

If you think this non-template-varargs code is fine to call `emplace()`, then sure, why not?

Reply:
My reason for wanting `emplace()` is that these datastructures look and feel like STL containers already, it would make sense to continue adopting those semantics. If I did not examine the headers carefully, I would first try and use `emplace()` and see that it doesn't compile, and then switch back to `insert(std::make_pair())` -- I would not expect `insert()` overload to accept two arguments.

Reply:
So you think it's better just to rename this to `emplace()`, even if I don't implement it "properly" with the varargs template stuff?  OK, no problem.

Reply:
Hmm. I've never attempted to implement `emplace()` before, but I'm guessing the templated varargs approach would be able to capture lvalue and rvalue without separate overloads?

Reply:
That's my understanding.